### PR TITLE
Feature | Add cart

### PR DESCRIPTION
**Pull request by WoLkErSs**

### CODE REVIEW

File: app/assets/stylesheets/cart.scss [hunk 1]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Cart controller here.
```
Comment (za4emnik) [inline app/assets/stylesheets/cart.scss]: remove

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -2,4 +2,34 @@ class ApplicationController < ActionController::Base
   protect_from_forgery with: :exception
   add_flash_types :success, :danger, :info, :warning
   helper :all
+  before_action :items_quantity
+
+  def items_quantity
```
Comment (za4emnik) [inline app/controllers/application_controller.rb]: move to helper

File: app/controllers/application_controller.rb [hunk 3]
```diff
@@ -2,4 +2,34 @@ class ApplicationController < ActionController::Base
...
+    @items_quantity = current_items_quantity
+  end
+
+  def current_order
+    Order.where(user_id: current_user.id).where(status: 'in_progress').first_or_create do |order|
```
Comment (za4emnik) [inline app/controllers/application_controller.rb]: .in_progress

File: app/controllers/application_controller.rb [hunk 4]
```diff
@@ -2,4 +2,34 @@ class ApplicationController < ActionController::Base
...
+      order.user_id = current_user.id
+    end
+  end
+
+  private
+
+  def current_items_quantity
```
Comment (za4emnik) [inline app/controllers/application_controller.rb]: separate on 2 methods

File: app/controllers/cart_controller.rb [hunk 5]
```diff
@@ -0,0 +1,60 @@
+class CartController < ApplicationController
+  attr_reader :items, :cart
+
+  ZERO = 0
+
+  def show
+    @items = GetOrderItemsService.new(current_user, session[:guest_items]).call
+    @cart = current_order if current_user
```
Comment (za4emnik) [inline app/controllers/cart_controller.rb]: change to order

File: app/controllers/cart_controller.rb [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    else
+      session[:guest_items][params[:book_id]] = params[:value]
+    end
+
+    redirect_to cart_path
+  end
+
+  def active_coupon
+    coupon = Coupon.find_by(key: params[:discount][:key], active: true, order_id: nil)
+    ImplementCouponServise.new(current_user).call(coupon, current_order) if current_user && coupon
```
Comment (za4emnik) [inline app/controllers/cart_controller.rb]: apply

File: app/controllers/cart_controller.rb [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+      add_item_to_order
+      UpdateOrderTotalService.new.call(current_order.id)
+    else
+      session[:guest_items][params[:book_id]] = params[:value]
+    end
+
+    redirect_to cart_path
+  end
+
+  def active_coupon
```
Comment (za4emnik) [inline app/controllers/cart_controller.rb]: move to coupons controller

File: app/controllers/cart_controller.rb [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    current_order.order_items.find_by(book_id: params[:book_id]).update(quantity: params[:value])
+  end
+
+  def setup_flash_coupon(coupon)
+    coupon ? flash[:notice] = t('cart.eneble_coupon') : flash[:danger] = t('cart.disabled_coupon')
+  end
+
+  def guest_total
+    coupon = Coupon.find_by(id: session[:coupon], active: true, order_id: nil)
+    coupon_value = coupon ? coupon.value : ZERO
```
Comment (za4emnik) [inline app/controllers/cart_controller.rb]: `coupon_value = coupon&.value || 0`

File: app/controllers/order_items_controller.rb [hunk 9]
```diff
@@ -8,29 +8,24 @@ def create
...
-
   def save_guest
-    session[:guest_items] ||= [] unless session[:guest_items]
-    session[:guest_items] << { id: params[:book_id], quantity: params[:quantity] }
-  end
-
-  def take_current_order
-    order = Order.find_by(user_id: current_user.id)
-    order || create_order
+    session[:guest_items] ||= {} unless session[:guest_items]
```
Comment (za4emnik) [inline app/controllers/order_items_controller.rb]: remove unless session[:guest_items]

File: app/services/create_order_item_service.rb [hunk 10]
```diff
@@ -1,5 +1,12 @@
 class CreateOrderItemService
-  def call(order, params)
-    OrderItem.create(order_id: order.id, book_id: params[:book_id].to_i, quantity: params[:quantity].to_i) if order
+  def call(order_id, params)
+    item = OrderItem.find_by(order_id: order_id, book_id: params[:book_id].to_i)
```
Comment (za4emnik) [inline app/services/create_order_item_service.rb]: remove to_i

File: app/services/create_order_item_service.rb [hunk 11]
```diff
@@ -1,5 +1,12 @@
...
+    if item
+      quantity = item.quantity
```
Comment (za4emnik) [inline app/services/create_order_item_service.rb]: separate

File: app/services/get_order_items_service.rb [hunk 12]
```diff
@@ -0,0 +1,37 @@
+class GetOrderItemsService
```
Comment (za4emnik) [inline app/services/get_order_items_service.rb]: rename please

File: app/services/get_order_items_service.rb [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+  def initialize(current_user, guest_items)
+    @current_order = current_user.orders.find_by(status: 'in_progress') if current_user
```
Comment (za4emnik) [inline app/services/get_order_items_service.rb]: .in_progress

File: app/services/get_order_items_service.rb [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+    order_items
+    @items
+  end
+
+  private
+
+  def order_items
+    if !@current_user && @guest_items
+      items_of_guest
+    elsif @current_user
```
Comment (za4emnik) [inline app/services/get_order_items_service.rb]: else

File: app/services/get_order_items_service.rb [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+      items_of_user
+    end
+  end
+
+  def items_of_user
+    @current_order.order_items.each do |item|
+      book = Book.find_by(id: item.book_id)
```
Comment (za4emnik) [inline app/services/get_order_items_service.rb]: item.book

File: app/services/get_order_items_service.rb [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+  def order_items
+    if !@current_user && @guest_items
+      items_of_guest
+    elsif @current_user
+      items_of_user
+    end
+  end
+
+  def items_of_user
+    @current_order.order_items.each do |item|
```
Comment (za4emnik) [inline app/services/get_order_items_service.rb]: order_items.includes(:book)

File: app/services/get_order_items_service.rb [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+      book = Book.find_by(id: item.book_id)
+      @items[book] = item.quantity
+    end
+  end
+
+  def items_of_guest
+    @guest_items.each do |book_id, quantity|
+      book = Book.find_by(id: book_id.to_i)
```
Comment (za4emnik) [inline app/services/get_order_items_service.rb]: `books = Book.where(id: [array of ids])`

File: app/services/implement_coupon_servise.rb [hunk 18]
```diff
@@ -0,0 +1,20 @@
+class ImplementCouponServise
```
Comment (za4emnik) [inline app/services/implement_coupon_servise.rb]: rename

File: app/services/implement_coupon_servise.rb [hunk 19]
```diff
@@ -0,0 +1,20 @@
...
+
+  private
+
+  def use_coupon(current_order, coupon)
+    update(current_order, coupon) unless current_order.coupon
+  end
+
+  def update(current_order, coupon)
+    coupon.update(order_id: current_order.id, active: false)
+    current_order.update(coupon_id: coupon.id)
```
Comment (za4emnik) [inline app/services/implement_coupon_servise.rb]: dont need

File: app/services/update_order_total_service.rb [hunk 20]
```diff
@@ -0,0 +1,17 @@
...
+  private
+
+  def update_total(order)
+    total = 0
+    order.order_items.each { |item| total += item.quantity * book_price(item.book_id) }
+    order.update(total: total)
+  end
+
+  def book_price(id)
+    Book.find_by(id: id).price
```
Comment (za4emnik) [inline app/services/update_order_total_service.rb]: the same

File: app/views/cart/show.haml [hunk 21]
```diff
@@ -0,0 +1,131 @@
...
+        %tr
+          %td
+            %span.in-grey-600 SubTotal
+          %td
+            %span.font-16.in-gold-500 €32.90
+    .general-cart-item.divider-lg-bottom.mb-30
+      %button.close.general-cart-close(aria-label="Close" type="button")
+        %span(aria-hidden="true") ×
+      .general-img-wrap-table
+        %img.cart-img-shadow(alt="design-book" src="http://jonathantweedy.com/resources/thumbs/SmashingBook5ResponsiveWebDesign.jpg")/
```
Comment (za4emnik) [inline app/views/cart/show.haml]: ?

File: app/views/cart/show.haml [hunk 22]
```diff
@@ -0,0 +1,131 @@
...
+              %td
+                %p.font-16
+                  = t(:'book.price', price: @cart.coupon.value) if current_user && @cart.coupon
+                  = t(:'book.price', price: @cart[:coupon_value]) unless current_user
+            %tr
+              %td
+                %strong.font-18 Order Total:
+              %td
+                %strong.font-18
+                  = t(:'book.price', price: ('%.2f' % (@cart.total - @cart.coupon.value))) if current_user && @cart.coupon
```
Comment (za4emnik) [inline app/views/cart/show.haml]: move to helper

File: spec/controllers/coupons_controller_spec.rb [hunk 23]
```diff
@@ -0,0 +1,146 @@
...
+      let(:coupon) { create(:coupon, :enebled_coupon, order: nil) }
+      let(:params) {{ discount: {key: coupon.key, value: coupon.value} }}
+
+      before { subject }
+
+      it 'save to session' do
+        expect(controller.session[:coupon]).to eq(coupon.id)
+      end
+
+      it 'alert applay coupon' do
```
Comment (za4emnik) [inline spec/controllers/coupons_controller_spec.rb]: apply

File: spec/controllers/coupons_controller_spec.rb [hunk 24]
```diff
@@ -0,0 +1,146 @@
...
+
+          it 'redirect to cart' do
+            expect(controller).to redirect_to(cart_path)
+          end
+        end
+      end
+
+      context 'when order total less then coupon' do
+
+        let(:coupon) { create(:coupon, :enebled_coupon, order: nil) }
```
Comment (za4emnik) [inline spec/controllers/coupons_controller_spec.rb]: enebled_coupon

File: spec/controllers/order_items_controller_spec.rb [hunk 25]
```diff
@@ -1,11 +1,135 @@
...
+      let(:quantity) { item.quantity + 1 }
+      let(:book_id) { item.book_id }
+      let(:params) { { 'book_id' => book_id, 'value' => quantity } }
+      context 'when update item' do
+        before do
+          login_user(user)
+        end
+
+        it 'update item quantity' do
+          expect { subject }.to change { OrderItem.find_by(book_id: book_id).quantity }.from(number).to(number+1)
```
Comment (za4emnik) [inline spec/controllers/order_items_controller_spec.rb]: .by(1)

File: spec/controllers/order_items_controller_spec.rb [hunk 26]
```diff
@@ -1,11 +1,135 @@
...
+
+      it 'have status 302' do
+        expect(response).to have_http_status(302)
+      end
+    end
+
+    context 'when user' do
+      let!(:user) { create(:user, :with_order) }
+      let(:item) { user.orders.in_progress.first.order_items.first }
+      let(:number) { item.quantity }
```
Comment (za4emnik) [inline spec/controllers/order_items_controller_spec.rb]: remove unneeded

File: spec/controllers/order_items_controller_spec.rb [hunk 27]
```diff
@@ -1,11 +1,135 @@
...
+      end
+    end
+  end
+
   describe '.create' do
+    let(:quantity) { '1' }
+    let(:order_item) { create(:order_item) }
+    let(:params) { { 'book_id' => order_item.id, 'quantity' => quantity } }
+
     subject(:controller_damn) { post :create, params: params }
```
Comment (za4emnik) [inline spec/controllers/order_items_controller_spec.rb]: controller_damn?

File: spec/controllers/orders_controller_spec.rb [hunk 28]
```diff
@@ -0,0 +1,4 @@
+require 'rails_helper'
+
+RSpec.describe OrdersController, type: :controller do
```
Comment (za4emnik) [inline spec/controllers/orders_controller_spec.rb]: remove

File: spec/decorators/book_decorator_spec.rb [hunk 29]
```diff
@@ -0,0 +1,21 @@
...
+
+  it 'returns the displayed days' do
+    expect(book.reviews_quantity).to eq(0)
+  end
+
+  it 'returns the displayed days' do
+    expect(book.authors_all).to eq(book.authors.map { |a| a.first_name + ' ' + a.last_name }.join(', '))
+  end
+
+  it 'returns the displayed days' do
```
Comment (za4emnik) [inline spec/decorators/book_decorator_spec.rb]: change name for `it`s

File: spec/factories/order_items.rb [hunk 30]
```diff
@@ -2,5 +2,6 @@
   factory :order_item do
     association :order
     association :book
+    quantity {5}
```
Comment (za4emnik) [inline spec/factories/order_items.rb]: rand(0..5)

File: spec/factories/orders.rb [hunk 31]
```diff
@@ -1,10 +1,17 @@
 FactoryBot.define do
   factory :order do
     association :user
+    total { 10 }
```
Comment (za4emnik) [inline spec/factories/orders.rb]: rand

File: spec/helpers/books_helper_spec.rb [hunk 32]
```diff
@@ -0,0 +1,18 @@
+require 'rails_helper'
```
Comment (za4emnik) [inline spec/helpers/books_helper_spec.rb]: remove

File: spec/helpers/cart_helper_spec.rb [hunk 33]
```diff
@@ -0,0 +1,20 @@
+require 'rails_helper'
+
+RSpec.describe CartHelper, type: :helper do
+  describe "#order" do
+    context 'when guest' do
+      it "returns the default value" do
+        expect(helper.order_summary).to eq({:coupon=>0, :subtotal=>0, :total=>0})
```
Comment (za4emnik) [inline spec/helpers/cart_helper_spec.rb]: remove hash rockets

File: spec/helpers/users_helper_spec.rb [hunk 34]
```diff
@@ -0,0 +1,15 @@
+require 'rails_helper'
+
+RSpec.describe UsersHelper, type: :helper do
+  describe "#billing_address" do
+      context 'should have addresses' do
+        it { expect(helper.billing_address).to be_kind_of(BillingForm) }
+        it { expect(helper.shipping_address).to be_kind_of(ShippingForm) }
+      end
+    # context 'when guest' do
```
Comment (za4emnik) [inline spec/helpers/users_helper_spec.rb]: ?

File: spec/queries/books_query_spec.rb [hunk 35]
```diff
@@ -0,0 +1,41 @@
...
+
+    it 'Sort price: High to low' do
+      expect(query.new(Book.all).books_sort('Price: High to low')).to eq([book_a, book_b, book_c])
+    end
+
+    it 'Default sort' do
+      expect(query.new(Book.all).books_sort('')).to eq([book_c, book_b, book_a])
+    end
+
+    it 'Default sort' do
```
Comment (za4emnik) [inline spec/queries/books_query_spec.rb]: name

File: spec/queries/books_query_spec.rb [hunk 36]
```diff
@@ -0,0 +1,41 @@
+require 'rails_helper'
+
+RSpec.describe BooksQuery do
+  subject(:query) { described_class }
+
+  let(:book_a) { create(:book, title: 'Aaaaa', price: 3, sold: 1) }
```
Comment (za4emnik) [inline spec/queries/books_query_spec.rb]: let!

File: spec/queries/books_query_spec.rb [hunk 37]
```diff
@@ -0,0 +1,41 @@
...
+  let(:book_b) { create(:book, title: 'Bbbbb', price: 2, sold: 5) }
+  let(:book_c) { create(:book, title: 'Ccccc', price: 1, sold: 3) }
+
+  context 'when update order item' do
+    before do
```
Comment (za4emnik) [inline spec/queries/books_query_spec.rb]: remove

File: spec/rails_helper.rb [hunk 38]
```diff
@@ -1,7 +1,26 @@
...
 ENV['RAILS_ENV'] ||= 'test'
+require 'simplecov'
+SimpleCov.start do
+  add_filter 'app/channels/application_cable/channel.rb'
+  add_filter 'app/channels/application_cable/connection.rb'
+  add_filter 'app/jobs/application_job.rb'
+  add_filter 'app/decorators/order_item_decorator.rb'
+  add_filter 'app/mailers/application_mailer.rb'
+  add_filter 'app/uploaders/image_uploader.rb'
+  add_filter 'app/services/utils/sekrets_service.rb'
```
Comment (za4emnik) [inline spec/rails_helper.rb]: 'app/admin/**/*'

File: spec/services/create_order_item_service_spec.rb [hunk 39]
```diff
@@ -0,0 +1,30 @@
...
+    end
+  end
+
+  context 'when create order item' do
+    let(:params) { { :book_id => order_item.book_id, :quantity => 1 } }
+
+    it 'update item' do
+      expect{ service.call(new_order.id, params) }.to change {
+        OrderItem.all.count
+      }.from(1).to(2)
```
Comment (za4emnik) [inline spec/services/create_order_item_service_spec.rb]: by(1)

File: spec/services/items_quantity_service_spec.rb [hunk 40]
```diff
@@ -0,0 +1,38 @@
...
+    let!(:order) { order_item.order }
+
+    it { expect( service.call(guest_items, user, order) ).to eq(1) }
+  end
+
+  context 'when add items to user' do
+    let(:guest_items) { {'1' => '1'} }
+    let!(:order_item) { create(:order_item) }
+    let!(:user) { order_item.order.user }
+    let!(:order) { order_item.order }
```
Comment (za4emnik) [inline spec/services/items_quantity_service_spec.rb]: move to top

File: spec/services/summary_service_spec.rb [hunk 41]
```diff
@@ -0,0 +1,45 @@
...
+RSpec.describe SummaryService do
+  let!(:book) { create(:book) }
+  let(:guest_items) { {} }
+  let(:coupon) { create(:coupon, order: nil) }
+
+  subject(:service) { described_class.new }
+
+  context 'when apply coupon which more when subtotal' do
+    it 'guest appla coupon' do
+      expect( service.call(nil, guest_items, coupon.id, nil) ).to eq({:coupon=>coupon.value, :subtotal=>0, :total=>0})
```
Comment (za4emnik) [inline spec/services/summary_service_spec.rb]: hash rockets

File: app/forms/address_form.rb [hunk 42]
```diff
@@ -9,20 +9,45 @@ class AddressForm
...
-  PHONE_LENGTH = 12
-  MAX_STRING = 50
+  VALIDATE_FORMAT = {
+    fields: /\A[a-zA-Z]*\z/,
+    address: /\A[a-zA-Z0-9 \-\,]*\z/,
+    zip: /\A[0-9\-]*\z/,
+    phone: /\A\+[0-9]*\z/
+  }.freeze
+
+  ZIP_LENGTH = 5
```
Comment (za4emnik) [inline app/forms/address_form.rb]: move

File: config/routes.rb [hunk 43]
```diff
@@ -5,16 +5,21 @@
...
+  resources :order_items, only: %i[create update destroy]
   resources :posts
-  resources :books, only: [:show, :index]
-  resources :reviews, only: [:create]
+  resources :books, only: %i[show index]
+  resources :reviews, only: %i[create]
   resources :pages
+  resources :orders, only: %i[show index]
+  resources :coupons, only: %i[update]
+  resource :cart, only: %i[show], controller: 'cart'
```
[the remark this answers is not in the document]
Comment (WoLkErSs) [reply config/routes.rb]: Fixed in festure/settings

**Pull request merged**